### PR TITLE
deps(react): Upgrade to 17

### PR DESCRIPTION
**Pull request by evanpurkhiser**
neat

This does not yet turn on the new react automatic runtime

### CODE REVIEW

**Comment by EvanPurkhiser** [comment]
Perhpase concerningly `tsc` now takes 20s more.

before
> Done in 71.05s.

after
> Done in 91.00s.

File: static/app/components/autoComplete.tsx [hunk 1]
```diff
@@ -31,19 +31,18 @@ const defaultProps = {
   shouldSelectWithTab: false,
 };
 
-type GetInputArgs = {
+type GetInputArgs<E extends HTMLInputElement> = {
```
Comment (EvanPurkhiser) [inline static/app/components/autoComplete.tsx:34]: I needed to make these generic to work around the fact that react's KeyboardEvent has become more tight as to interoperability between Element types.

File: tests/js/setup.js [hunk 2]
```diff
@@ -26,6 +26,11 @@ if (!SVGElement.prototype.getTotalLength) {
 
 /**
  * Enzyme configuration
+ *
+ * TODO(epurkhiser): We're using @wojtekmaj's react-17 enzyme adapter, until
+ * the offical adapter has been released.
+ *
+ * https://github.com/enzymejs/enzyme/issues/2429
```
Comment (markstory) [inline tests/js/setup.js:33]: Is enzyme often lagging behind react something we should be taking into consideration as a reason to consider React Testing Library more? From [this comment](https://github.com/enzymejs/enzyme/issues/2429#issuecomment-817750268) it sounds like react-testing-library doesn't use react internals and is easier to upgrade with.

Comment (billyvg) [reply tests/js/setup.js:33]: I think this is a pretty strong argument for adopting React Testing Library

File: package.json [hunk 3]
```diff
@@ -149,13 +149,13 @@
     "@storybook/theming": "^6.2.8",
     "@visual-snapshot/jest": "^2.0.2",
     "@webpack-cli/serve": "^1.3.0",
+    "@wojtekmaj/enzyme-adapter-react-17": "^0.6.1",
```
Comment (billyvg) [inline package.json:152]: Can we hard pin this at `0.6.1` since it's a rando fork?

Comment (EvanPurkhiser) [reply package.json:152]: Yeah

Comment (EvanPurkhiser) [reply package.json]: done

**Comment by EvanPurkhiser** [comment]
OK I re-aligned the versions in the lock file so that we weren't installing two separate packages, that seems to have helped 🤔

**Review by billyvg — APPROVED** [review]

File: tests/js/setup.js (hunk 2 above)
Comment (priscilawebdev) [reply tests/js/setup.js:33]: Agree!!!!

**Pull request merged**